_linalg_qr_helper meta function

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #78836
* #78601
* #78604
* #78527
* __->__ #78603

Taken from https://github.com/albanD/subclass_zoo/blob/main/python_meta_tensor.py

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
_linalg_qr_helper meta function

Taken from https://github.com/albanD/subclass_zoo/blob/main/python_meta_tensor.py

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

lezcano reviewed (approved):
Just one check missing, otherwise lgtm

Lezcano commented on torch/_meta_registrations.py:
```diff
@@ -212,3 +212,34 @@ def meta_repeat_interleave_Tensor(repeats, output_size=None):
     return repeats.new_empty(output_size)
+
+@torch.library.impl(meta_lib, "_linalg_qr_helper")
+def meta_linalg_qr_helper(input, mode):
```
It's actually also missing the chekc on the dtype of `input`.

ezyang replied:
it doesn't look like the og function checks these

```
std::tuple<Tensor, Tensor> _linalg_qr_helper_default(const Tensor& input, c10::string_view mode) {
  bool compute_q, reduced_mode;
  std::tie(compute_q, reduced_mode) = _parse_qr_mode(mode);
  auto m = input.size(-2);
  auto n = input.size(-1);
  auto mn = std::min(m, n);

  // Allocate Q, R tensors with correct shape and memory layout
  Tensor Q;
  if (compute_q) {
    auto Qt_shape = input.sizes().vec();
    Qt_shape.end()[-2] = reduced_mode ? mn : m;
    Qt_shape.end()[-1] = m;
    Q = at::empty(Qt_shape, input.options());
    Q.transpose_(-2, -1); // make 'Q' with Fortran contiguous memory layout
  } else {
    Q = at::empty({0}, input.options());
  }

  auto Rt_shape = input.sizes().vec();
  Rt_shape.end()[-2] = n;
  Rt_shape.end()[-1] = (reduced_mode || !compute_q) ? mn : m;
  Tensor R = at::empty(Rt_shape, input.options());
  R.transpose_(-2, -1); // make 'R' with Fortran contiguous memory layout

  // Now fill Q, R tensors with the result
  linalg_qr_out_helper(input, Q, R, compute_q, reduced_mode);

  return std::make_tuple(Q, R);
}

```

ezyang replied:
alright i'm wrong it's in the out helper lol

Lezcano replied:
The preparation code for this function could be heavily simplified with Yukio's recent additions to the structured kernels logic. I should port it to those at some point.

ezyang committed:
Update on "_linalg_qr_helper meta function"


Taken from https://github.com/albanD/subclass_zoo/blob/main/python_meta_tensor.py

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "_linalg_qr_helper meta function"


Taken from https://github.com/albanD/subclass_zoo/blob/main/python_meta_tensor.py

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

mruberry reviewed (approved):
Stamped

ezyang commented on torch/_meta_registrations.py:
```diff
@@ -212,3 +212,34 @@ def meta_repeat_interleave_Tensor(repeats, output_size=None):
+    n = input.size(-1)
+    mn = min(m, n)
+    if compute_q:
+        Qt_shape = list(input.size())
+        Qt_shape[-2] = mn if reduced_mode else m
+        Qt_shape[-1] = m
```
don't feel like golfing lol

ezyang committed:
Update on "_linalg_qr_helper meta function"


Taken from https://github.com/albanD/subclass_zoo/blob/main/python_meta_tensor.py

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "_linalg_qr_helper meta function"


Taken from https://github.com/albanD/subclass_zoo/blob/main/python_meta_tensor.py

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "_linalg_qr_helper meta function"


Taken from https://github.com/albanD/subclass_zoo/blob/main/python_meta_tensor.py

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "_linalg_qr_helper meta function"


Taken from https://github.com/albanD/subclass_zoo/blob/main/python_meta_tensor.py

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "_linalg_qr_helper meta function"


Taken from https://github.com/albanD/subclass_zoo/blob/main/python_meta_tensor.py

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge
